### PR TITLE
Improve evaluation of nested `ComposedFunction`s

### LINKED ISSUE #45877: Deeply composed functions much slower than equivalent static functions
```julia
julia> VERSION
v"1.9.0-DEV.869"

julia> eval(:(f(x) = $(reduce((s,x) -> :($x($s)), fill(sin,500); init = :x))))
f (generic function with 1 method)
# f(x) = sin(sin(...(sin(x))))

julia> g = reduce(∘, fill(sin,500));
# g(x) = (sin∘sin∘...∘sin)(x)

julia> @time f(1.0) == g(1.0)
  0.686770 seconds (1.66 M allocations: 109.801 MiB, 5.39% gc time, 98.75% compilation time)
true

julia> @time f(1.0)
  0.000001 seconds
0.07698641344541407

julia> @time g(1.0)
  0.011081 seconds (1000 allocations: 15.625 KiB)
0.07698641344541407
```
The compile time is vastly improved since 1.7, but the evaluation performance is worse
```julia
julia> VERSION
v"1.7.3"

julia> eval(:(f(x) = $(reduce((s,x) -> :($x($s)), fill(sin,500); init = :x))))
f (generic function with 1 method)

julia> g = reduce(∘, fill(sin,500));

julia> @time f(1.0) == g(1.0)
114.266682 seconds (228.16 M allocations: 13.787 GiB, 2.16% gc time, 100.00% compilation time)
true

julia> @time f(1.0)
  0.000010 seconds
0.07698641344541407

julia> @time g(1.0)
  0.000244 seconds (499 allocations: 7.797 KiB)
0.07698641344541407
```

### PR DESCRIPTION
**Pull request by martinholters**
* ~~Add (the equivalent of) `@assume_effects :terminates_globally` to `unwrap_composed`. Although it could be inferred as `Const`, without the annotation, it was not elided for too complex inputs, resulting in unnecessary runtime overhead.~~
  EDIT: now #45993 is merged and this part isn't necessary.
* Reverse recursion order in `call_composed`. This prevents potentially changing argument types being piped through the recursion, making inference bail out. With the reversed order, only the tuple of remaining functions is changing during recursion and is becoming strictly simpler, letting inference succeed.

```julia
julia> f = reduce(∘, fill(sin,500));

julia> const g = reduce(∘, fill(sin,500));

julia> @time f(1.0)
  1.474751 seconds (1.83 M allocations: 120.273 MiB, 3.22% gc time, 99.00% compilation time) # master
  6.011575 seconds (4.02 M allocations: 278.036 MiB, 3.33% gc time, 100.00% compilation time) # PR
0.07698641344541407

julia> @btime f(1.0)
  14.102 ms (1000 allocations: 15.62 KiB) # master
  5.337 μs (1 allocation: 16 bytes) # PR
0.07698641344541407

julia> @btime g(1.0)
  14.091 ms (998 allocations: 15.59 KiB) # master
  1.878 ns (0 allocations: 0 bytes) # PR
0.07698641344541407
```
So increase in compilation time is noticeable, but execution is also much faster. (Yes, we can const-prop and completely elide `g` here). Also note that compilation time is still much better than with 1.8-rc1 (or 1.7), where it is several minutes.

That said, the fact that we have to replace the concise `(c::ComposedFunction)(x...; kw...) = c.outer(c.inner(x...; kw...))` with the convoluted construction still makes me sad.

Fixes #45877.

### CODE REVIEW

File: base/operators.jl
Comment (N5N3) [inline base/operators.jl:1022]: There seems no need to split `fs[1]` with reversed recursion.
```julia
(c::ComposedFunction)(x...; kw...) = call_composed(unwrap_composed(c), x, kw)
```

**Review by N5N3 — APPROVED** [review]

File: base/operators.jl
Comment (aviatesk) [inline base/operators.jl:1032]: I think we generally want to avoid direct creation of `:purity` expression in this way as it is fragile against future changes in the effect system design.

I'd like to add this diff:
```diff
diff --git a/base/essentials.jl b/base/essentials.jl
index 906c36ad9c..8eee2923b3 100644
--- a/base/essentials.jl
+++ b/base/essentials.jl
@@ -204,23 +204,35 @@ macro _pure_meta()
 end
 # can be used in place of `@assume_effects :total` (supposed to be used for bootstrapping)
 macro _total_meta()
-    return _is_internal(__module__) && Expr(:meta, Expr(:purity,
-        #=:consistent=#true,
-        #=:effect_free=#true,
-        #=:nothrow=#true,
-        #=:terminates_globally=#true,
-        #=:terminates_locally=#false,
-        #=:notaskstate=#true))
+    return _is_internal(__module__) && Expr(:meta, _assumed_effects(;
+        consistent=true,
+        effect_free=true,
+        nothrow=true,
+        terminates_globally=true,
+        notaskstate=true))
 end
 # can be used in place of `@assume_effects :foldable` (supposed to be used for bootstrapping)
 macro _foldable_meta()
-    return _is_internal(__module__) && Expr(:meta, Expr(:purity,
-        #=:consistent=#true,
-        #=:effect_free=#true,
-        #=:nothrow=#false,
-        #=:terminates_globally=#true,
-        #=:terminates_locally=#false,
-        #=:notaskstate=#false))
+    return _is_internal(__module__) && Expr(:meta, _assumed_effects(;
+        consistent=true,
+        effect_free=true,
+        terminates_globally=true))
+end
+
+function _assumed_effects(;
+    consistent::Bool = false,
+    effects_free::Bool = false,
+    nothrow::Bool = false,
+    terminates_globally::Bool = false,
+    terminates_locally::Bool = false,
+    notaskstate::Bool = false)
+    return Expr(:purity,
+        consistent,
+        effects_free,
+        nothrow,
+        terminates_globally,
+        terminates_locally,
+        notaskstate)
 end
 
 # another version of inlining that propagates an inbounds context
```

and then rewrite this piece of code like:
```suggestion
eval(quote
    # equivalent to `@assume_effects :terminates_globally`,
    # but @assume_effects cannot be used `Core.Compiler`, hence `_assumed_effects(...)`
    function unwrap_composed(c::ComposedFunction)
        $(Expr(:meta, _assumed_effects(; terminates_globally=true)))
        return (unwrap_composed(c.outer)..., unwrap_composed(c.inner)...)
    end
end)
```

Then we don't look for all the usages of `:purity` expression when we add new effect to the system as far as we keep the same naming schema for the existing effects.

Comment (aviatesk) [reply base/operators.jl:1032]: It seems more difficult than I expected to use kwfunc until we define namedtuple.jl.. We can leave this to the future and don't need to let it hold this PR.

**Comment by aviatesk** [comment]
You can test this improvement as like:
> test/operators.jl
```julia
include("compiler/irutils.jl")
const issue_45877 = reduce(∘, fill(sin,500))
@test fully_eliminated() do
    issue_45877(1.0)
end
```

File: base/operators.jl
Comment (martinholters) [reply base/operators.jl]: Alternatively, we could move the `ComposedFunction` stuff to a dedicated file that is included after `array.jl` (which makes `@assume_effects` work) but before `bitarray.jl` (which uses composition with `!iszero`).

**Comment by martinholters** [comment]
@aviatesk can you explain why the `@assume_effects` is necessary? On master I find
```julia
julia> const g = reduce(∘, fill(sin,500));

julia> only(code_typed() do
           Base.unwrap_composed(g)
       end)[2]
NTuple{500, typeof(sin)}

julia> Base.infer_effects() do
           Base.unwrap_composed(g)
       end
(+c,+e,+n,!t,+s)
```
So inference succeeds which implies that it follows the recursion to the end, therefore obviously has figured out that there is an end. But the effects don't seem to catch that. Is there some way we could make the effects here more precise? If we could avoid the `@assume_effects`, there would also no longer be a need for the code movement.

**Review by aviatesk — APPROVED** [review]
We can try to improve the recursion detection for effect analysis instead, but I'm also fine with merging this for now and tackle the improvement later.

**Comment by martinholters** [comment]
I don't mind either way. I won't have the bandwidth to look at improving effect analysis anytime soon, don't know whether you can tackle it in the foreseeable future. If no, we probably should merge as is. But I have no personal interest in deeply nested composed functions, so no hurry from my side.

**Comment by ianatol** [comment]
LGTM.

I think the `@assume_effects` is fine for the time being, while Shuhei looks into improving the recursion detection for effects.

**Pull request merged**
